sopel/irc/modes.py

```diff
@@ -0,0 +1,234 @@
+        raise ModeTypeUnknown(mode)
+
+    def get_mode_info(
+        self,
+        mode: str,
+        is_added: bool
+    ) -> Tuple[Optional[str], bool, bool]:
+        """Retrieve ``mode``'s information when added or removed.
+
+        :raise ModeTypeUnknown: when the mode's type is unknown and isn't a
+                                user privilege
+        :raise ModeTypeImproperlyConfigured: when the mode's type is known but
+                                             there is no information for
+                                             parameters (if and when they are
+                                             required by the mode)
+        :return: a tuple with three values: the mode type, if it requires a
+                 parameter, and if it's a channel mode or a user privilege
+
+        ::
+
+            >>> chanmodes = {'A': tuple('beI'), 'B': tuple('k')}
+            >>> t_params = {'A': PARAM_ALWAYS, 'B': PARAM_ADDED}
+            >>> mm = ModeMessage(chanmodes, t_params)
+            >>> mm.get_mode_info('e', False)
+            ('A', True, False)
+            >>> mm.get_mode_info('k', False)
+            ('B', False, False)
+            >>> mm.get_mode_info('v', True)
+            (None, True, True)
+
+        .. note::
+
+            A user privilege ``mode`` doesn't have a type so the first value
+            returned will be ``None`` in that case.
+        """
+        try:
+            letter = self.get_mode_type(mode)
+        except ModeTypeUnknown:
+            if mode in self.privileges:
+                # a user privilege doesn't have a type
+                return None, True, True
+            # not a user privilege: re-raise error
+            raise
+
+        if letter not in self.type_params:
+            # we don't know how to handle this type of mode
+            raise ModeTypeImproperlyConfigured(mode, letter)
+
+        type_param = self.type_params[letter]
+        return letter, not type_param == PARAM_NEVER and (
+            type_param == PARAM_ALWAYS
+            or type_param == PARAM_ADDED and is_added
+            or type_param == PARAM_REMOVED and not is_added
+        ), False
+
+    def parse_modestring(
+        self,
+        modestring: str,
+        params: Tuple[str, ...]
+    ) -> ModeString:
+        """Parse a ``modestring`` for a channel with its ``params``.
+
+        :return: the parsed and validated information for that ``modestring``
+        """
+        imodes = iter(_modes_is_added(modestring))
+        iparams = iter(params)
+        modes: List = []
+        privileges: List = []
+        for mode, is_added in imodes:
+            param = None
+            try:
+                letter, required, is_priv = self.get_mode_info(mode, is_added)
+                if required:
+                    try:
+                        param = next(iparams)
+                    except StopIteration:
+                        # Not enough parameters: we have to stop here
```

Comment:
Should log the event.

Having params left over after consuming all mode characters should also log that event.

Reply:
If you look at how it is used, there *are* logs. Just not here where we parse everything and return a result with all the information (valid modes, leftover parameters, missing parameters, etc.).